src/client_side_reply.cc

```diff
@@ -2140,7 +2140,8 @@ clientReplyContext::createStoreEntry(const HttpRequestMethod& m, RequestFlags re
      */
 
     if (http->request == NULL) {
-        const MasterXaction::Pointer mx = new MasterXaction(XactionInitiator::initClient);
+        const auto connManager = http->getConn();
+        const auto mx = MasterXaction::MakePortful(connManager ? connManager->port : nullptr);
```

Comment:
That a "Portful" xaction can be passed nullptr is a bug.

This should be:
```suggestion
        MasterXaction::Pointer mx;
        if (const auto connManager = http->getConn())        
            mx = MasterXaction::MakePortful(connManager->port);
        else
            mx = MasterXaction::MakePortless(XactionInitiator::initClient);
```

or
```suggestion
        const auto mx = MasterXaction::MakePortful(XactionInitiator::initClient);
```

Reply:
The meaning of the words "Portful" and "Portless" are opposites and mutually exclusive. It is a semantic and logical flaw for **both** to mean "no port exists". Otherwise there is no point in having the "Portless" variant at all.

There were two suggestions, I have now corrected the typo in second. Though I think the first is what we all prefer.

Reply:
Note that both your suggestions have _MakePortless(XactionInitiator::initClient)_, leading to a static assertion.

Reply:
> The meaning of the words "Portful" and "Portless" are opposites and mutually exclusive.

Correct.

> It is a semantic and logical flaw for both to mean "no port exists" (the case where nullptr is accepted by Portful).

You are criticizing a meaning that does not exist. Portful means "the port should exist (but, due to problems beyond caller control, information about the port may have been lost)". The existing method descriptions attempt to document this, but we should improve that documentation if it is not clear enough. Please suggest such improvements.

```diff
-        const auto mx = MasterXaction::MakePortful(connManager ? connManager->port : nullptr);
+        const auto mx = MasterXaction::MakePortless(XactionInitiator::initClient);
```

The whole point of these methods is to prevent `MakePortless(initClient)` or similar calls (at _compile_ time) because humans will place them where they should not exist -- bugs notwithstanding, all initClient contexts are associated with a port.

Squid code does have bugs, and we cannot fix them all in this PR. Thus, we want to emphasize that something went wrong in a particular context by explicitly supplying a nil port to MakePortful. The contradiction is fairly obvious and that is the whole point of this design. There is no way to guarantee the absence of bugs, of course, but the proposed way attracts attention to code that requires special attention. Like the code this change request is attached to.

That special attention does not imply that the caller code itself is necessarily buggy. It only means that the author and reviewer must make sure the port information may have been lost indeed, for reasons (i.e. bugs) outside the caller control.

Reply:
You are both nitpicking the specific use of ```initClient``` in a **suggestion**.

You missed that my point was about using an if-statement so that ```MakePortful()``` is **only** called when a port is known. ```MakePortless()``` whenever it is unknown or lost.


> 
> The whole point of these methods is to prevent `MakePortless(initClient)` or similar calls (at _compile_ time) because humans will place them where they should not exist -- bugs notwithstanding, all initClient contexts are associated with a port.

... except when they are associated with a nullptr. Like this PR line does.

To me accepting nulltpr into MakePortful() is one of those very bugs this PR change is trying to catch.

i.e.
This MakePortful() accepting nullptr is a clear demonstration that devs can/should use ```MakePortful(foo, nullptr)``` any time they want.
 -> In other words; your own code demonstrates that MakePortless() is pointless fluff which can be dropped.
 -> Or to take it one step further; demonstrating that the original code we have right now is good enough and no benefit is added by this PR.

I think we can agree that those implications are bad and we should not to accept code with the pattern ```MakePortful(nullptr)```

Reply:
I am not nitpicking at all. I am trying to explain what this PR is about because the suggestions in this change request have missed the very essence of this PR.  AFAICT, Eduard is doing the same, but he is using compiler output -- a static assertion -- to illustrate essentially the same point.

To make progress, we should first try to reach an understanding of what problem this PR is trying to address. Only after we develop such a shared understanding it makes sense to discuss specific solutions to that problem. Currently, it looks like this precondition has not been met yet.

The problem this PR is trying address is folks creating a MasterXaction object with an initClient initiator while _forgetting_ to supply the port information to that object, even though that port information _is_ available. We cannot guarantee that such bugs will never happen, but we can provide an API that decreases the probability of such bugs.

@yadij, does my explanation clarify what risk this PR tries to reduce? Do you agree that the risk of forgetting to supply the port (when port is available) is worth decreasing? If yes, then we can discuss the best way to reduce that risk.

> To me accepting nulltpr into MakePortful() is one of those very bugs this PR change is trying to catch.

Exactly! Every MakePortful(nullptr) is a bug. This PR claims that all similar expressions in this PR are old bugs that we cannot easily fix (and should not fix) in this PR. The PR essentially marks those old bugs. There is no harm in such markings and there is some obvious benefit, but those old bugs are not what this PR is focusing on.

The key idea behind this PR is to decrease the probability of new/future bugs of the same kind -- forgetting to supply an available port. Every time an author or reviewer sees a new MakePortful(nullptr) call being added to Squid, they should pause -- "What is going on here? Did I/they forget to supply the required port information?". Hopefully, we will answer such questions correctly in most cases and reduce the number of _new_ bugs that way.

Eventually, all old bugs will be fixed as well, and we will be able to prohibit nil ports for initClient MasterXactions. It may take a while though because weak pointers that lose port information are widespread in Squid.

If we allow MakePortful(initClient) or MakePortful() expressions, then the idea behind this PR stops working. Authors and reviewers are a lot less likely to think twice about those expressions because they look "normal" or "ordinary". It is the explicit `nullptr` in `MakePortful(nullptr)` that makes (some of) us pause and think.

Reply:
> @yadij, does my explanation clarify what risk this PR tries to reduce?

Yes, and I do understand that.

What I am trying to get across is that ...

> > To me accepting nulltpr into MakePortful() is one of those very bugs this PR change is trying to catch.
> 
> Exactly! Every MakePortful(nullptr) is a bug. This PR claims that all similar expressions in this PR are old bugs that we cannot easily fix (and should not fix) in this PR. The PR essentially marks those old bugs. There is no harm in such markings and there is some obvious benefit, but those old bugs are not what this PR is focusing on.
> 

IMO **for this PRs case** simply replacing the API and leaving this bug is not good enough ...


> If we allow MakePortful(initClient) or MakePortful() expressions, then the idea behind this PR stops working. Authors and reviewers are a lot less likely to think twice about those expressions because they look "normal" or "ordinary". It is the explicit `nullptr` in `MakePortful(nullptr)` that makes (some of) us pause and think.

I am okay with ```MakePortless(initClient)``` being prohibited.

What I am against is ```MakePortful(nulltpr)``` being left as allowed. Specifically because this Downloader usage is essentially **the** bug with current API. This PR having a seamless change of all other code demonstrates that.

Reply:
Maybe an alternative suggestion would help clarify that I don't really care **how** this bug is removed, just that it is ...

```suggestion
        const auto connManager = http->getConn();
        assert(connManager); // XXX: Downloader without any client connection not yet supported
        const auto mx = MasterXaction::MakePortful(connManager->port);
```

Reply:
@eduard-bagdasaryan, can we fix this particular out of scope bug if we require all clientReplyContext creators to supply their MasterXaction object (that the clientReplyContext constructor will store via a shared pointer)? In other words, refactor clientReplyContext to never create a new MasterXaction object?

@eduard-bagdasaryan, I believe the official code creating initClient MasterXaction here is another out-of-scope bug. Some of these transactions are not necessarily initClient AFAICT. Do you agree?

Reply:
> Some of these transactions are not necessarily initClient AFAICT

I think you are right. For example, clientReplyContext created from clientBeginRequest() initiated by ESIInclude::Start() should have initEsi instead.

Reply:
Sounds good. Please try implementing that. If that fix works, I hope we will be able to require port information for all initClient master transactions.